File: app/src/main/java/com/snap/lighting/LightSceneRenderer.java
package com.snap.lighting;

import android.opengl.GLES20;
import android.opengl.GLSurfaceView;
import android.opengl.Matrix;

import com.snap.model.GlHelpers;
import com.snap.model.exception.GlLibException;

import java.nio.FloatBuffer;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

public class LightSceneRenderer implements GLSurfaceView.Renderer {

    private float xСameraPosition, yCameraPosition, zCameraPosition;

    private float xLightPosition, yLightPosition, zLightPosition;

    private final float[] modelMatrix = new float[16];
    private final float[] viewMatrix = new float[16];
    private final float[] modelViewMatrix = new float[16];
    private final float[] projectionMatrix = new float[16];
    private final float[] modelViewProjectionMatrix = new float[16];

    private FloatBuffer mSeaVerticesBuffer;
    private FloatBuffer mSkyVerticesBuffer;
    private FloatBuffer mMainSailVerticesBuffer;
    private FloatBuffer mSmallSailVerticesBuffer;
    private FloatBuffer mBoatVerticesBuffer;

    private FloatBuffer mVerticesNormalsBuffer;

    private FloatBuffer mSeaVerticesColorsBuffer;
    private FloatBuffer mSkyVerticesColorsBuffer;
    private FloatBuffer mAnySailVerticesColorsBuffer;
    private FloatBuffer mBoatVerticesColorsBuffer;

    private LightSceneShadingProgram mSeaShader;
    private LightSceneShadingProgram mSkyShader;
    private LightSceneShadingProgram mMainSailShader;
    private LightSceneShadingProgram mSmallSailShader;
    private LightSceneShadingProgram mBoatShader;


    public LightSceneRenderer() {
        setup();
    }

    private void setup() {
        setupLightSource();
        setupModelViewMatrix();
        setupVertexBuffers();
        setupNormalsBuffer();
        setupVerticesColorBuffers();
    }

    /**
     * Point source of light.
     */
    private void setupLightSource() {
        xLightPosition = 0.5f;
        yLightPosition = 0.2f;
        zLightPosition = 0.5f;
    }

    private void setupModelViewMatrix() {
        //мы не будем двигать объекты поэтому сбрасываем модельную матрицу на единичную
        Matrix.setIdentityM(modelMatrix, 0);

        //координаты камеры
        xСameraPosition = 0.0f;
        yCameraPosition = 0.0f;
        zCameraPosition = 3.0f;

        // пусть камера смотрит на начало координат
        // и верх у камеры будет вдоль оси Y
        // зная координаты камеры получаем матрицу вида
        Matrix.setLookAtM(
                viewMatrix, 0, xСameraPosition, yCameraPosition, zCameraPosition,
                0, 0, 0, 0, 1, 0);
        // умножая матрицу вида на матрицу модели
        // получаем матрицу модели-вида
        Matrix.multiplyMM(modelViewMatrix, 0, viewMatrix, 0, modelMatrix, 0);
    }

    private void setupVertexBuffers() {
        float[] seaVertices = {
                -1.0f, -0.35f, 0.0f,
                -1.0f, -1.5f, 0.0f,
                1.0f, -0.35f, 0.0f,
                1.0f, -1.5f, 0.0f
        };
        float[] skyVertices = {
                -1.0f, 1.5f, 0.0f,
                -1.0f, -0.35f, 0.0f,
                1.0f, 1.5f, 0.0f,
                1.0f, -0.35f, 0
        };
        float[] mainSailVertices = {
                -0.5f, -0.45f, 0.4f,
                0.0f, -0.45f, 0.4f,
                0.0f, 0.5f, 0.4f
        };
        float[] smallSailVertices = {
                0.05f, -0.45f, 0.4f,
                0.22f, -0.5f, 0.4f,
                0.0f, 0.25f, 0.4f
        };
        float[] boatVertices = {
                -0.5f, -0.5f, 0.4f,
                -0.5f, -0.6f, 0.4f,
                0.22f, -0.5f, 0.4f,
                0.18f, -0.6f, 0.4f
        };

        mSeaVerticesBuffer = GlHelpers.createNativeFloatBuffer(seaVertices);
        mSkyVerticesBuffer = GlHelpers.createNativeFloatBuffer(skyVertices);
        mMainSailVerticesBuffer = GlHelpers.createNativeFloatBuffer(mainSailVertices);
        mSmallSailVerticesBuffer = GlHelpers.createNativeFloatBuffer(smallSailVertices);
        mBoatVerticesBuffer = GlHelpers.createNativeFloatBuffer(boatVertices);
    }

    private void setupNormalsBuffer() {
        //вектор нормали перпендикулярен плоскости квадрата
        //и направлен вдоль оси Z
        float nx = 0;
        float ny = 0;
        float nz = 1;
        //нормаль одинакова для всех вершин квадрата,
        //поэтому переписываем координаты вектора нормали в массив 4 раза
        float normalArray[] = {
                nx, ny, nz,
                nx, ny, nz,
                nx, ny, nz,
                nx, ny, nz
        };
        mVerticesNormalsBuffer = GlHelpers.createNativeFloatBuffer(normalArray);
    }

    private void setupVerticesColorBuffers() {
        // R-G-B-A
        float[] seaVerticesColors = {
                0f, 1f, 1f, 1,
                0f, 0f, 1f, 1,
                0f, 1f, 1f, 1,
                0f, 0f, 1f, 1,
        };
        float[] skyVerticesColors = {
                0.2f, 0.2f, 0.8f, 1,
                0.5f, 0.5f, 1, 1,
                0.2f, 0.2f, 0.8f, 1,
                0.5f, 0.5f, 1, 1,
        };
        float[] sailVerticesColors = {
                1, 0.1f, 0.1f, 1,
                1, 1, 1, 1,
                1, 0.1f, 0.1f, 1,
        };
        float[] boatVerticesColors = {
                1, 1, 1, 1,
                0.2f, 0.2f, 0.2f, 1,
                1, 1, 1, 1,
                0.2f, 0.2f, 0.2f, 1,
        };
        mSeaVerticesColorsBuffer = GlHelpers.createNativeFloatBuffer(seaVerticesColors);
        mSkyVerticesColorsBuffer = GlHelpers.createNativeFloatBuffer(skyVerticesColors);
        mAnySailVerticesColorsBuffer = GlHelpers.createNativeFloatBuffer(sailVerticesColors);
        mBoatVerticesColorsBuffer = GlHelpers.createNativeFloatBuffer(boatVerticesColors);
    }

    /**
     * Here we are going to setup Projection and Model-View-Projection matrices.
     */
    public void onSurfaceChanged(GL10 unused, int width, int height) {
        // устанавливаем glViewport
        GLES20.glViewport(0, 0, width, height);

        float ratio = (float) width / height;
        float k = 0.047f;
        float left = -k * ratio;
        float right = k * ratio;
        float bottom = -k;
        float top = k;
        float near = 0.1f;
        float far = 10.0f;

        Matrix.frustumM(projectionMatrix, 0, left, right, bottom, top, near, far);
        Matrix.multiplyMM(modelViewProjectionMatrix,
                0, projectionMatrix,
                0, modelViewMatrix,
                0);
    }

    public void onSurfaceCreated(GL10 unused, EGLConfig config) {
        //включаем тест глубины
        GLES20.glEnable(GLES20.GL_DEPTH_TEST);
        //включаем отсечение невидимых граней
        GLES20.glEnable(GLES20.GL_CULL_FACE);
        //включаем сглаживание текстур, это пригодится в будущем
        GLES20.glHint(GLES20.GL_GENERATE_MIPMAP_HINT, GLES20.GL_NICEST);

        mSeaShader = LightSceneShadingProgram.newInstance();
        mSkyShader = LightSceneShadingProgram.newInstance();
        mMainSailShader = LightSceneShadingProgram.newInstance();
        mSmallSailShader = LightSceneShadingProgram.newInstance();
        mBoatShader = LightSceneShadingProgram.newInstance();

        try {
            mSeaShader.setup();
            mSkyShader.setup();
            mMainSailShader.setup();
            mSmallSailShader.setup();
            mBoatShader.setup();
        } catch (GlLibException e) {
            throw new RuntimeException("LightSceneShadingProgram setting up error. Reason: " + e);
        }
    }

    /**
     * Frame rendering
     */
    public void onDrawFrame(GL10 unused) {
        // clean frame
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT | GLES20.GL_DEPTH_BUFFER_BIT);

        // Render Sea
        linkAttributesAndUniforms(mSeaShader,
                mSeaVerticesBuffer, mVerticesNormalsBuffer, mSeaVerticesColorsBuffer);
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, GlHelpers.BYTES_PER_FLOAT);

        // Render Sky
        linkAttributesAndUniforms(mSkyShader,
                mSkyVerticesBuffer, mVerticesNormalsBuffer, mSkyVerticesColorsBuffer);
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, GlHelpers.BYTES_PER_FLOAT);

        // Render MainSail
        linkAttributesAndUniforms(mMainSailShader,
                mMainSailVerticesBuffer, mVerticesNormalsBuffer, mAnySailVerticesColorsBuffer);
        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 3);

        // Render SmallSail
        linkAttributesAndUniforms(mSmallSailShader,
                mSmallSailVerticesBuffer, mVerticesNormalsBuffer, mAnySailVerticesColorsBuffer);
        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 3);

        // Render Boat
        linkAttributesAndUniforms(mBoatShader,
                mBoatVerticesBuffer, mVerticesNormalsBuffer, mBoatVerticesColorsBuffer);
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, GlHelpers.BYTES_PER_FLOAT);
    }

    private void linkAttributesAndUniforms(LightSceneShadingProgram shadingProgram,
                                           FloatBuffer verticesBuffer,
                                           FloatBuffer verticesNormalsBuffer,
                                           FloatBuffer verticesColorsBuffer) {

        shadingProgram.linkVertexBuffer(verticesBuffer);
        shadingProgram.linkNormalBuffer(verticesNormalsBuffer);
        shadingProgram.linkColorBuffer(verticesColorsBuffer);

        shadingProgram.linkModelViewProjectionMatrix(modelViewProjectionMatrix);
        shadingProgram.linkCamera(xСameraPosition, yCameraPosition, zCameraPosition);
        shadingProgram.linkLightSource(xLightPosition, yLightPosition, zLightPosition);
    }

}

File: app/src/main/java/com/snap/lighting/LightSceneRenderer.kt
package com.snap.lighting

import android.opengl.GLES20
import android.opengl.GLSurfaceView
import android.opengl.Matrix


import java.nio.FloatBuffer

import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.opengles.GL10

import com.snap.model.bytesPerFloat
import com.snap.model.createNativeFloatBuffer

class LightSceneRenderer : GLSurfaceView.Renderer {

    private var xСameraPosition: Float = 0.toFloat()
    private var yCameraPosition: Float = 0.toFloat()
    private var zCameraPosition: Float = 0.toFloat()

    private var xLightPosition: Float = 0.toFloat()
    private var yLightPosition: Float = 0.toFloat()
    private var zLightPosition: Float = 0.toFloat()

    private val modelMatrix = FloatArray(16)
    private val viewMatrix = FloatArray(16)
    private val modelViewMatrix = FloatArray(16)
    private val projectionMatrix = FloatArray(16)
    private val modelViewProjectionMatrix = FloatArray(16)

    private lateinit var seaVerticesBuffer: FloatBuffer
    private lateinit var skyVerticesBuffer: FloatBuffer
    private lateinit var mainSailVerticesBuffer: FloatBuffer
    private lateinit var smallSailVerticesBuffer: FloatBuffer
    private lateinit var boatVerticesBuffer: FloatBuffer

    private lateinit var verticesNormalsBuffer: FloatBuffer

    private lateinit var seaVerticesColorsBuffer: FloatBuffer
    private lateinit var skyVerticesColorsBuffer: FloatBuffer
    private lateinit var anySailVerticesColorsBuffer: FloatBuffer
    private lateinit var boatVerticesColorsBuffer: FloatBuffer

    private lateinit var seaShader: LightSceneShadingProgram
    private lateinit var skyShader: LightSceneShadingProgram
    private lateinit var mainSailShader: LightSceneShadingProgram
    private lateinit var smallSailShader: LightSceneShadingProgram
    private lateinit var boatShader: LightSceneShadingProgram

    init {
        setup()
    }

    private fun setup() {
        setupLightSource()
        setupModelViewMatrix()
        setupVertexBuffers()
        setupNormalsBuffer()
        setupVerticesColorBuffers()
    }

    /**
     * Point source of light.
     */
    private fun setupLightSource() {
        xLightPosition = 0.5f
        yLightPosition = 0.2f
        zLightPosition = 0.5f
    }

    private fun setupModelViewMatrix() {
        //мы не будем двигать объекты поэтому сбрасываем модельную матрицу на единичную
        Matrix.setIdentityM(modelMatrix, 0)

        //координаты камеры
        xСameraPosition = 0.0f
        yCameraPosition = 0.0f
        zCameraPosition = 3.0f

        // пусть камера смотрит на начало координат
        // и верх у камеры будет вдоль оси Y
        // зная координаты камеры получаем матрицу вида
        Matrix.setLookAtM(
                viewMatrix, 0, xСameraPosition, yCameraPosition, zCameraPosition,
                0f, 0f, 0f, 0f, 1f, 0f)
        // умножая матрицу вида на матрицу модели
        // получаем матрицу модели-вида
        Matrix.multiplyMM(modelViewMatrix, 0, viewMatrix, 0, modelMatrix, 0)
    }

    private fun setupVertexBuffers() {
        val seaVertices = floatArrayOf(-1.0f, -0.35f, 0.0f, -1.0f, -1.5f, 0.0f, 1.0f, -0.35f, 0.0f, 1.0f, -1.5f, 0.0f)
        val skyVertices = floatArrayOf(-1.0f, 1.5f, 0.0f, -1.0f, -0.35f, 0.0f, 1.0f, 1.5f, 0.0f, 1.0f, -0.35f, 0f)
        val mainSailVertices = floatArrayOf(-0.5f, -0.45f, 0.4f, 0.0f, -0.45f, 0.4f, 0.0f, 0.5f, 0.4f)
        val smallSailVertices = floatArrayOf(0.05f, -0.45f, 0.4f, 0.22f, -0.5f, 0.4f, 0.0f, 0.25f, 0.4f)
        val boatVertices = floatArrayOf(-0.5f, -0.5f, 0.4f, -0.5f, -0.6f, 0.4f, 0.22f, -0.5f, 0.4f, 0.18f, -0.6f, 0.4f)

        seaVerticesBuffer = createNativeFloatBuffer(seaVertices)
        skyVerticesBuffer = createNativeFloatBuffer(skyVertices)
        mainSailVerticesBuffer = createNativeFloatBuffer(mainSailVertices)
        smallSailVerticesBuffer = createNativeFloatBuffer(smallSailVertices)
        boatVerticesBuffer = createNativeFloatBuffer(boatVertices)
    }

    private fun setupNormalsBuffer() {
        //вектор нормали перпендикулярен плоскости квадрата
        //и направлен вдоль оси Z
        val nx = 0f
        val ny = 0f
        val nz = 1f
        //нормаль одинакова для всех вершин квадрата,
        //поэтому переписываем координаты вектора нормали в массив 4 раза
        val normalArray = floatArrayOf(nx, ny, nz, nx, ny, nz, nx, ny, nz, nx, ny, nz)
        verticesNormalsBuffer = createNativeFloatBuffer(normalArray)
    }

    private fun setupVerticesColorBuffers() {
        // R-G-B-A
        val seaVerticesColors = floatArrayOf(0f, 1f, 1f, 1f, 0f, 0f, 1f, 1f, 0f, 1f, 1f, 1f, 0f, 0f, 1f, 1f)
        val skyVerticesColors = floatArrayOf(0.2f, 0.2f, 0.8f, 1f, 0.5f, 0.5f, 1f, 1f, 0.2f, 0.2f, 0.8f, 1f, 0.5f, 0.5f, 1f, 1f)
        val sailVerticesColors = floatArrayOf(1f, 0.1f, 0.1f, 1f, 1f, 1f, 1f, 1f, 1f, 0.1f, 0.1f, 1f)
        val boatVerticesColors = floatArrayOf(1f, 1f, 1f, 1f, 0.2f, 0.2f, 0.2f, 1f, 1f, 1f, 1f, 1f, 0.2f, 0.2f, 0.2f, 1f)
        seaVerticesColorsBuffer = createNativeFloatBuffer(seaVerticesColors)
        skyVerticesColorsBuffer = createNativeFloatBuffer(skyVerticesColors)
        anySailVerticesColorsBuffer = createNativeFloatBuffer(sailVerticesColors)
        boatVerticesColorsBuffer = createNativeFloatBuffer(boatVerticesColors)
    }

    /**
     * Here we are going to createAndCompile Projection and Model-View-Projection matrices.
     */
    override fun onSurfaceChanged(unused: GL10, width: Int, height: Int) {
        // устанавливаем glViewport
        GLES20.glViewport(0, 0, width, height)

        val ratio = width.toFloat() / height
        val k = 0.047f
        val left = -k * ratio
        val right = k * ratio
        val bottom = -k
        val near = 0.1f
        val far = 10.0f

        Matrix.frustumM(projectionMatrix, 0, left, right, bottom, k, near, far)
        Matrix.multiplyMM(modelViewProjectionMatrix,
                0, projectionMatrix,
                0, modelViewMatrix,
                0)
    }

    override fun onSurfaceCreated(unused: GL10, config: EGLConfig) {
        //включаем тест глубины
        GLES20.glEnable(GLES20.GL_DEPTH_TEST)
        //включаем отсечение невидимых граней
        GLES20.glEnable(GLES20.GL_CULL_FACE)
        //включаем сглаживание текстур, это пригодится в будущем
        GLES20.glHint(GLES20.GL_GENERATE_MIPMAP_HINT, GLES20.GL_NICEST)

        seaShader = LightSceneShadingProgram.newInstance()
        skyShader = LightSceneShadingProgram.newInstance()
        mainSailShader = LightSceneShadingProgram.newInstance()
        smallSailShader = LightSceneShadingProgram.newInstance()
        boatShader = LightSceneShadingProgram.newInstance()

        listOf(seaShader, skyShader, mainSailShader, smallSailShader, boatShader).forEach { it.setup() }
    }

    /**
     * Frame rendering
     */
    override fun onDrawFrame(unused: GL10) {
        // clean frame
        GLES20.glClear(GLES20.GL_COLOR_BUFFER_BIT or GLES20.GL_DEPTH_BUFFER_BIT)

        // Render Sea
        linkAttributesAndUniforms(seaShader,
                seaVerticesBuffer, verticesNormalsBuffer, seaVerticesColorsBuffer)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, bytesPerFloat)

        // Render Sky
        linkAttributesAndUniforms(skyShader,
                skyVerticesBuffer, verticesNormalsBuffer, skyVerticesColorsBuffer)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, bytesPerFloat)

        // Render MainSail
        linkAttributesAndUniforms(mainSailShader,
                mainSailVerticesBuffer, verticesNormalsBuffer, anySailVerticesColorsBuffer)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 3)

        // Render SmallSail
        linkAttributesAndUniforms(smallSailShader,
                smallSailVerticesBuffer, verticesNormalsBuffer, anySailVerticesColorsBuffer)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLES, 0, 3)

        // Render Boat
        linkAttributesAndUniforms(boatShader,
                boatVerticesBuffer, verticesNormalsBuffer, boatVerticesColorsBuffer)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, bytesPerFloat)
    }

    private fun linkAttributesAndUniforms(shadingProgram: LightSceneShadingProgram,
                                          verticesBuffer: FloatBuffer,
                                          verticesNormalsBuffer: FloatBuffer,
                                          verticesColorsBuffer: FloatBuffer) {

        shadingProgram.linkVertexBuffer(verticesBuffer)
        shadingProgram.linkNormalBuffer(verticesNormalsBuffer)
        shadingProgram.linkColorBuffer(verticesColorsBuffer)

        shadingProgram.linkModelViewProjectionMatrix(modelViewProjectionMatrix)
        shadingProgram.linkCamera(xСameraPosition, yCameraPosition, zCameraPosition)
        shadingProgram.linkLightSource(xLightPosition, yLightPosition, zLightPosition)
    }

}